[Bug]: BigQueryIO DIRECT_READ breaks for empty table during Bounded-to-Unbounded conversion

### What happened?

## Summary

When using `BigQueryIO.read().withMethod(Method.DIRECT_READ)` to read from an **empty BigQuery table** or a table that returns zero rows after applying `RowRestriction`, the BigQuery Storage API server returns a `ReadSession` with **zero streams**.  `BigQueryStorageSourceBase.split()` handles this by returning `ImmutableList.of()` (an empty list of sources). 
While this works correctly for purely bounded reads, it **breaks any pipeline that wraps this bounded source into an unbounded one** (via `UnboundedReadFromBoundedSource`).
When `split()` returns an empty list, `UnboundedReadFromBoundedSource` falls back to wrapping the original unsplit source directly (`ImmutableList.of(boundedSource)`). However, `BigQueryStorageSourceBase.createReader()` is not implemented — it unconditionally throws `UnsupportedOperationException("BigQuery storage source must be split before reading")`, because it is designed to only be read through its per-stream sub-sources (`BigQueryStorageStreamSource`). This causes the pipeline to get stuck in a loop of exceptions.

## Environment

- **Apache Beam SDK version:** 2.70.0 (reproduced; likely affects all versions with DIRECT_READ support)
- **Runner:** Dataflow Runner v1

## Steps to Reproduce

1. Create a Beam pipeline that uses `BigQueryIO.read().withMethod(Method.DIRECT_READ)` to read from a BigQuery table.
2. Use this read in a streaming pipeline.
3. Ensure the target BigQuery table is **empty**, or returns zero rows after `RowRestriction` is applied.
4. Run the pipeline.

## Observed Behavior

- The pipeline **gets stuck** and the following error appears repeatedly in logs:

  ```
  Caused by: java.lang.UnsupportedOperationException: BigQuery storage source must be split before reading
      org.apache.beam.sdk.io.gcp.bigquery.BigQueryStorageSourceBase.createReader(BigQueryStorageSourceBase.java:200)
      org.apache.beam.sdk.io.gcp.bigquery.BigQueryStorageTableSource.createReader(BigQueryStorageTableSource.java:42)
      org.apache.beam.sdk.util.construction.UnboundedReadFromBoundedSource$BoundedToUnboundedSourceAdapter$ResidualSource.advance(UnboundedReadFromBoundedSource.java:497)
  ```

## Expected Behavior

The pipeline does not stall — downstream operations proceed normally, simply processing zero elements from this branch.

## Additional Context

### Chain of events

1. `BigQueryIO.read().withMethod(Method.DIRECT_READ)` uses `BigQueryStorageSourceBase` as the underlying `BoundedSource`.
2. `split()` builds a `CreateReadSessionRequest` and sends it to the server.
3. The server returns a `ReadSession` with `streams_count = 0` for an empty table.
4. `split()` returns `ImmutableList.of()`.
5. `UnboundedReadFromBoundedSource$BoundedToUnboundedSourceAdapter` receives an empty list of sub-sources and falls back to wrapping the original unsplit source directly (`ImmutableList.of(boundedSource)`). When it calls `createReader()` on this `BigQueryStorageSourceBase`, the method throws `UnsupportedOperationException("BigQuery storage source must be split before reading")` because `BigQueryStorageSourceBase` does not implement `createReader()` — it is designed to only be read through its per-stream sub-sources.
6. The job gets stuck in a loop, repeatedly hitting this exception without making progress. The watermark never advances and the pipeline must be manually cancelled.


### Issue Priority

Priority: 2 (default / most bugs should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [x] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [x] Component: IO connector
- [ ] Component: Beam YAML
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Infrastructure
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [x] Component: Google Cloud Dataflow Runner

--- a/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageSourceBase.java
+++ b/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageSourceBase.java
@@ -30,6 +30,7 @@ import com.google.cloud.bigquery.storage.v1.ReadSession;
 import com.google.cloud.bigquery.storage.v1.ReadStream;
 import java.io.IOException;
 import java.util.List;
+import java.util.NoSuchElementException;
 import org.apache.beam.sdk.coders.Coder;
 import org.apache.beam.sdk.io.BoundedSource;
 import org.apache.beam.sdk.io.gcp.bigquery.BigQueryServices.StorageClient;
@@ -41,6 +42,7 @@ import org.apache.beam.sdk.util.Preconditions;
 import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.ImmutableList;
 import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.Lists;
 import org.checkerframework.checker.nullness.qual.Nullable;
+import org.joda.time.Instant;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -203,7 +205,52 @@ abstract class BigQueryStorageSourceBase<T> extends BoundedSource<T> {
 
   @Override
   public BoundedReader<T> createReader(PipelineOptions options) throws IOException {
+    try {
+      if (split(0, options).isEmpty()) {
+        return new EmptyReader<>(this);
+      }
+    } catch (Exception e) {
+      // If split fails, we can't be sure if it's empty or not.
+      // For backwards compatibility with tests that don't mock everything,
+      // we still throw UnsupportedOperationException.
+    }
     throw new UnsupportedOperationException("BigQuery storage source must be split before reading");
+  }
+
+  private static class EmptyReader<T> extends BoundedReader<T> {
+    private final BigQueryStorageSourceBase<T> source;
+
+    EmptyReader(BigQueryStorageSourceBase<T> source) {
+      this.source = source;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      return false;
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      return false;
+    }
+
+    @Override
+    public T getCurrent() throws NoSuchElementException {
+      throw new NoSuchElementException();
+    }
+
+    @Override
+    public Instant getCurrentTimestamp() throws NoSuchElementException {
+      throw new NoSuchElementException();
+    }
+
+    @Override
+    public void close() throws IOException {}
+
+    @Override
+    public BoundedSource<T> getCurrentSource() {
+      return source;
+    }
   }
 
   private void setPicosTimestampPrecision(

--- a/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageSourceBase.java
+++ b/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageSourceBase.java
@@ -213,6 +213,7 @@ abstract class BigQueryStorageSourceBase<T> extends BoundedSource<T> {
       // If split fails, we can't be sure if it's empty or not.
       // For backwards compatibility with tests that don't mock everything,
       // we still throw UnsupportedOperationException.
+      LOG.debug("Split failed during createReader emptiness check", e);
     }
     throw new UnsupportedOperationException("BigQuery storage source must be split before reading");
   }

--- a/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageSourceBase.java
+++ b/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageSourceBase.java
@@ -74,6 +74,7 @@ abstract class BigQueryStorageSourceBase<T> extends BoundedSource<T> {
   protected final Coder<T> outputCoder;
   protected final BigQueryServices bqServices;
   private final @Nullable TimestampPrecision picosTimestampPrecision;
+  private boolean emptyOrPruned = false;
 
   BigQueryStorageSourceBase(
       @Nullable DataFormat format,
@@ -181,8 +182,10 @@ abstract class BigQueryStorageSourceBase<T> extends BoundedSource<T> {
     if (readSession.getStreamsList().isEmpty()) {
       LOG.info(
           "Returned stream list is empty. The underlying table is empty or all rows have been pruned.");
+      emptyOrPruned = true;
       return ImmutableList.of();
     } else {
+      emptyOrPruned = false;
       LOG.info("Read session returned {} streams", readSession.getStreamsList().size());
     }
 
@@ -205,15 +208,11 @@ abstract class BigQueryStorageSourceBase<T> extends BoundedSource<T> {
 
   @Override
   public BoundedReader<T> createReader(PipelineOptions options) throws IOException {
-    try {
-      if (split(0, options).isEmpty()) {
-        return new EmptyReader<>(this);
-      }
-    } catch (Exception e) {
-      // If split fails, we can't be sure if it's empty or not.
-      // For backwards compatibility with tests that don't mock everything,
-      // we still throw UnsupportedOperationException.
-      LOG.debug("Split failed during createReader emptiness check", e);
+    if (emptyOrPruned) {
+      // When split() returns an empty list, UnboundedReadFromBoundedSource falls back to wrapping
+      // the original unsplit source directly (ImmutableList.of(bigQuerySotrageSourceBase)) so we
+      // need to return empty reader.
+      return new EmptyReader<>(this);
     }
     throw new UnsupportedOperationException("BigQuery storage source must be split before reading");
   }

--- a/sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOStorageReadTest.java
+++ b/sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOStorageReadTest.java
@@ -99,6 +99,7 @@ import org.apache.beam.sdk.extensions.protobuf.ByteStringCoder;
 import org.apache.beam.sdk.extensions.protobuf.ProtoCoder;
 import org.apache.beam.sdk.io.BoundedSource;
 import org.apache.beam.sdk.io.BoundedSource.BoundedReader;
+import org.apache.beam.sdk.io.UnboundedSource;
 import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO.TableRowParser;
 import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO.TypedRead;
 import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO.TypedRead.Method;
@@ -120,6 +121,7 @@ import org.apache.beam.sdk.testing.TestPipeline;
 import org.apache.beam.sdk.transforms.PTransform;
 import org.apache.beam.sdk.transforms.SerializableFunction;
 import org.apache.beam.sdk.transforms.display.DisplayData;
+import org.apache.beam.sdk.util.construction.UnboundedReadFromBoundedSource;
 import org.apache.beam.sdk.values.KV;
 import org.apache.beam.sdk.values.PBegin;
 import org.apache.beam.sdk.values.PCollection;
@@ -703,6 +705,49 @@ public class BigQueryIOStorageReadTest {
     thrown.expect(UnsupportedOperationException.class);
     thrown.expectMessage("BigQuery storage source must be split before reading");
     tableSource.createReader(options);
+  }
+
+  @Test
+  public void testUnboundedReadFromBoundedSourceWithEmptyTable() throws Exception {
+    fakeDatasetService.createDataset("project-id", "dataset", "", "", null);
+    TableReference tableRef = BigQueryHelpers.parseTableSpec("project-id:dataset.table");
+
+    Table table =
+        new Table().setTableReference(tableRef).setNumBytes(0L).setSchema(new TableSchema());
+
+    fakeDatasetService.createTable(table);
+
+    ReadSession emptyReadSession = ReadSession.newBuilder().build();
+    StorageClient fakeStorageClient = mock(StorageClient.class);
+    when(fakeStorageClient.createReadSession(any())).thenReturn(emptyReadSession);
+
+    BigQueryStorageTableSource<TableRow> tableSource =
+        BigQueryStorageTableSource.create(
+            ValueProvider.StaticValueProvider.of(tableRef),
+            null,
+            null,
+            new TableRowParser(),
+            TableRowJsonCoder.of(),
+            new FakeBigQueryServices()
+                .withDatasetService(fakeDatasetService)
+                .withStorageClient(fakeStorageClient));
+
+    // This simulates what happens in a streaming pipeline when BoundedSource is used
+    UnboundedSource<TableRow, ?> unboundedSource =
+        new UnboundedReadFromBoundedSource.BoundedToUnboundedSourceAdapter<>(tableSource);
+
+    // Initial split
+    List<? extends UnboundedSource<TableRow, ?>> splits = unboundedSource.split(1, options);
+    // Because tableSource.split returns empty list, BoundedToUnboundedSourceAdapter falls back to
+    // returning itself
+    assertEquals(1, splits.size());
+    UnboundedSource<TableRow, ?> splitSource = splits.get(0);
+
+    // Create reader
+    UnboundedSource.UnboundedReader<TableRow> reader = splitSource.createReader(options, null);
+
+    // This should NOT throw UnsupportedOperationException
+    assertFalse(reader.start());
   }
 
   private static GenericRecord createRecord(String name, Schema schema) {

